File: random_bench.go
package main

import (
	"fmt"
	"math/rand"
	"time"

	"golang.org/x/text/language"
	"golang.org/x/text/message"
)

func main() {
	rewards := []int{1, 2, 3, 4}

	length := len(rewards)

	totalPercent := float64(sum(rewards))

	countById := make([]int64, len(rewards))

	runCount := 100000000

	start := time.Now()

	for j := 0; j < runCount; j++ {
		random := rand.Float64() * totalPercent
		for i := 0; i < length; i++ {
			random -= float64(rewards[i])
			if random <= 0 {
				countById[i]++
				break
			}
		}
	}

	elapsed := time.Since(start)

	printer := message.NewPrinter(language.English)

	fmt.Printf("Total run: %s, elapsed: %.2fs. %d.00 ns/run, %s runs/s\n", printer.Sprintf("%d", runCount), elapsed.Seconds(), (elapsed / time.Duration(runCount)).Nanoseconds(), printer.Sprintf("%d", int64(runCount)/elapsed.Milliseconds()*1000))

	for i, count := range countById {
		fmt.Printf("id: %d count: %d (%.6f%%)\n", i, count, float64(count)/float64(runCount)*100)
	}
}

func sum(slice []int) int {
	s := 0
	for _, v := range slice {
		s += v
	}
	return s
}

File: random_bench.js
function main() {
    const rewards = [1, 2, 3, 4];

    const length = rewards.length;

    const totalPercent = rewards.reduce((sum, reward) => sum + reward, 0);

    const countById = new Array(rewards.length).fill(0);

    const runCount = 100000000;

    const start = new Date();

    for (let j = 0; j < runCount; j++) {
        let random = Math.random() * totalPercent;
        for (let i = 0; i < length; i++) {
            random -= rewards[i];
            if (random <= 0) {
                countById[i]++;
                break;
            }
        }
    }

    const elapsed = new Date() - start;

    console.log(
        `Total run: ${numberWithCommas(runCount)}, elapsed: ${(elapsed / 1000).toFixed(2)}s. ${(elapsed / runCount * 1e6
        ).toFixed(2)} ns/run, ${numberWithCommas(Math.floor(
            (runCount / (elapsed / 1000))
        ))} runs/s`
    );

    countById.forEach((count, i) => {
        console.log(`id: ${i}, count: ${count} (${((count / runCount) * 100).toFixed(6)}%)`);
    });
}

function numberWithCommas(x) {
    return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

main();
